src/main/resources/static/js/custom/validator.js

```diff
@@ -0,0 +1,128 @@
+import { DynamicProperty } from "./dynamic-property.js";
+import { Constants } from "./global.js";
+
+export class Validator {
+
+    /** @type {Validator} */
+    static #instance;
+
+    /** @type {Validator} */
+    static get instance() {
+        return Validator.#instance ??= new Validator(Constants.HtmlElements.FORM_ELEMENT);
+    }
+
+    /** @type {HTMLFormElement} */
+    form;
+
+     /**
+      * The validation roster contains all the validations that need to be performed before the form is submitted.
+      * They are organised in a Map by HTMLElement (using the element's id as a key). 
+      * For each HTMLElement we maintain a list of the dynamic-properties that need to be validated.
+      * 
+      * There are two reasons we need to keep a roster:
+      * 1. So that we can trigger all validations without dispatching "change" events.
+      * 2. So that we can remove event handlers after an HTMLElement has been removed from the DOM.
+      * 
+      *  @type {Map<String, DynamicProperty[]>} 
+      */
+    validationRoster = new Map();
+
+    /**
+     * 
+     * @param {String} formElementId 
+     */
+    constructor(formElementId) {
+
+        if (Validator.#instance) {
+            throw new Error("You should not instantiate the Validator class yourself. It is mean to be used as singleton.")
+        }
+
+        this.form = document.getElementById(formElementId);
+        if (this.form === null) {
+            throw new Error(`Form element "${formElementId}" not found.`)
+        }
+
+        this.form.setAttribute("novalidate", "");
+        this.form.addEventListener("submit", Validator.onSubmit);
+    }
+
+    checkValidity() {
+        for(const elementId of this.validationRoster.keys()) {
+            this.checkValidityOfElement(document.getElementById(elementId))
+        }
+    }
+
+    checkValidityOfElement(htmlElement) {
+        if (htmlElement.readOnly || htmlElement.disabled) {
+            return;
+        }
+
+        for (const property of this.validationRoster.get(htmlElement.id)) {
+            property.checkValidity();
```

Comment:
property.checkValidity()
property.htmlElement.checkValidity();
I have not read the entire code but it looks like the first it could do both in one call.

Reply:
It could but we do not want it to. do that. I was considering to use a different name for the DependencyProperty.checkvalidity method (like defineValidity) but I have left it like this until I get a better idea.